### PR TITLE
Do not leak SkipFrame exception to parent frames

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91059

Discovered by https://github.com/pytorch/pytorch/issues/93486, we noticed the exception `SkipFrame` to avoid repeatedly compiling frame of loop with graph breaks could leak to parent frames while inlining, which then prevents compiling.

This PR checks at inlining if such exception is raised and would instead raise an `Unsupported` to the outer frame. The original behavior and goal of #88857 is unaffected: the inner frame that has loop would still be skipped. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by BowenBao** [comment]
Inductor test failure should be unrelated. Segfaults locally on `test_torch.py -k test_scalar_check_cuda` without changes from this PR.

**Review by jansel — APPROVED** [review]

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -2065,6 +2065,56 @@ def fn(x):
...
+                i -= 1
+                torch._dynamo.graph_break()
+            return x
+
+        cnt = torch._dynamo.testing.CompileCounter()
+        opt_fn = torch._dynamo.optimize(cnt)(fn)
+        x = torch.randn(4)
+        opt_fn(x)
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(cnt.op_count, 1)
```
Comment (thiagocrepaldi) [inline test/dynamo/test_repros.py:2094]: This test is so useful and common we could have a helper for it

```python
def assert_no_graph_break(fn, inputs):
    cnt = torch._dynamo.testing.CompileCounter()
    opt_fn = torch._dynamo.optimize(cnt)(fn)
    opt_fn(inputs)
    self.assertEqual(cnt.frame_count, 1)
    self.assertEqual(cnt.op_count, 1)
```

**Review by thiagocrepaldi — APPROVED** [review]

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (jansel) [reply test/dynamo/test_repros.py:2094]: Agreed that would be a good refactor

Comment (BowenBao) [reply test/dynamo/test_repros.py:2094]: Agree this is nice thing to do. I (or anyone interested) could do it in a separate PR.

**Pull request closed without merging**